Listener: Add listener filters.

jrajahalme opened the pull request:
Network::ListenerImpl class implements an libevent socket
listener. Factor out all functionality that is not directly libevent
related to ConnectionHandlerImpl, which both reduces overall
complexity and makes it possible to refactor configurable listener
behavior to a new class of listener filters in the following patches.

This patch also reduces the special handling for SSL connections by
introducing a new dispatcher member for creating server
connections. This is used by the refactored code as the connection
creation is now shifted from Network::Listener to the connection
handler.

To further simplify the resulting code, this patch removes
Network::ListenerOptions class, as the only option that now needs to
be passed to Network::Listener is the "bind to port" boolean, which is
now passed in explicitly.  All other listener options are handled by
passing along Server::Listener instead.

Add support for listener filters and refactor listener original dst
and proxy protocol features as listener filters.

Existing use_proxy_protocol and use_original_dst listener flags are
still supported by internally converting them to listener filters.
    
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

Risk Level: High
Testing: Existing unit tests are converted to use the new filter framework where applicable.
Docs Changes: TBD
Release Notes: TBD
API Changes: [Data Plane PR #400][https://github.com/envoyproxy/data-plane-api/pull/400]
Deprecated: TBD if the legacy flags should be removed from v2 API

mattklein123 commented:
@PiotrSikora FYI this will have implications for the changes we were discussing today. I would wait for this to land first, otherwise you will need to rebase again. I will review this tomorrow.

PiotrSikora commented:
@lizan Could you take a look and see if this conflicts or helps with your work on transport security?

jrajahalme commented:
> On Jan 10, 2018, at 6:34 PM, Piotr Sikora <notifications@github.com> wrote:
> 
> @lizan <https://github.com/lizan> Could you take a look and see if this conflicts or helps with your work on transport security?
> 
> 

I briefly looked into the possibility of implementing SSL as a listener filter. IMO it seems possible using https://github.com/libevent/libevent/blob/master/bufferevent_openssl.c <https://github.com/libevent/libevent/blob/master/bufferevent_openssl.c> as long as Buffers will be evbuffers.

  Jarno

mattklein123 commented:
> I briefly looked into the possibility of implementing SSL as a listener filter. IMO it seems possible using https://github.com/libevent/libevent/blob/master/bufferevent_openssl.c <https://github.com/libevent/libevent/blob/master/bufferevent_openssl.c> as long as Buffers will be evbuffers.

We won't be going back to bufferevent, but @lizan is working on a bunch of refactoring around pluggable transport layers, and @PiotrSikora is working on sniffing TLS so that we can do filter chain matching based on sniffed data. (Basically have a listener support both plain-text and TLS). Then I'm working on v2 filter chain matching.

We just need to be careful here to line everything up. I'm hopeful we can sort it out in comments but it's possible we will need to have a quick meeting. Let's see how things go in comments tomorrow.

lizan commented:
Took a rough pass on the PR, some of this PR (e.g. deleting addSslListener from dispatcher) will help me for upstream transport socket, and I don't think it really add difficulty. One thing that needs clarification is when a listener filter read/write from/to the fd, it might conflict with what the transport socket do later.

Though this certainly conflicts with my local working directory. I'd like to know how long this is going to take to merge, to better lining things up.

mattklein123 commented:
@lizan I would like to move this along since it needs to come in front of a bunch of other items we are all working on. I've gotten derailed by a few things today so reviewing this might not happen but I will review later or tomorrow. We will aim to iterate and merge this early next week.

lizan commented:
@mattklein123 SGTM, I'll hold my changes for a few days :)

jrajahalme commented:
I still have some fixing left due to the rebase, will sort out tomorrow. Should not block reviewing, though.

mattklein123 commented on include/envoy/event/dispatcher.h:
```diff
@@ -38,6 +38,16 @@ class Dispatcher {
+   * @param accept_socket supplies a socket with an open file descriptor and connection metadata
+   *        to use for the connection. Takes ownership of the accept_socket.
+   * @param ssl_ctx supplies the SSL context to use, if not nullptr.
+   * @return Network::ConnectionPtr a client connection that is owned by the caller.
```
"a server connection"

jrajahalme replied:
Fixed!

mattklein123 commented on include/envoy/event/dispatcher.h:
```diff
@@ -38,6 +38,16 @@ class Dispatcher {
+   * @param ssl_ctx supplies the SSL context to use, if not nullptr.
+   * @return Network::ConnectionPtr a client connection that is owned by the caller.
+   */
+  virtual Network::ConnectionPtr createConnection(Network::AcceptSocketPtr&& accept_socket,
```
I would probably call this `createServerConnection` to match `createClientConnection`

jrajahalme replied:
Will change.

jrajahalme replied:
There is no ServerConnection class, ClientConnection inherits from Connection, so from that viewpoint createConnection() is correct. Hosts also have a createConnection method, so createServerConnection makes grepping the source easier.

mattklein123 commented on include/envoy/event/dispatcher.h:
```diff
@@ -85,25 +95,9 @@ class Dispatcher {
-  createSslListener(Network::ConnectionHandler& conn_handler, Ssl::ServerContext& ssl_ctx,
-                    Network::ListenSocket& socket, Network::ListenerCallbacks& cb,
-                    Stats::Scope& scope, const Network::ListenerOptions& listener_options) PURE;
+  virtual Network::ListenerPtr createListener(Network::ListenSocket& socket,
```
please fix doc comments above

jrajahalme replied:
OK, sorry for the sloppiness...

mattklein123 commented on include/envoy/network/connection_handler.h:
```diff
@@ -10,6 +10,11 @@
+namespace Server {
+class Listener;
+}
+// XXX: Move to Server namespace?
```
yeah, from an abstraction level in general we try to make server code depend on common code, but not the other way around. Why was this necessary?

jrajahalme replied:
Listener config is defined in the Server namespace (Server::Listener), and the ConnectionHandler's addListener() now takes it as an argument. I did not want to include server headers from here, so I used a bare class declaration (in the correct namespace) to circumvent the need for the include.

Could also see if the Listener config class could be moved to the common code.

mattklein123 replied:
Yeah, let's just move Server::Listener (that interface only) into the Network namespace. That interface doesn't depend on anything else in Server. Please do that in a dedicated PR since it will be just code movement. We can review/merge that quickly and then rebase this.

jrajahalme replied:
Network namespace already has a Listener (the abstract base class, so something needs to be renamed. How about renaming the Server::Listener as Network::ListenerConfig, as it is just that, "A configuration for an individual listener", as the comment says?

mattklein123 replied:
+1

jrajahalme replied:
Will do this and the filter directory split tomorrow, hopefully. In the meanwhile I updated this PR so that you may continue the review as you please :-)

mattklein123 commented on include/envoy/network/filter.h:
```diff
@@ -147,6 +148,67 @@ class FilterManager {
+  virtual Event::Dispatcher& dispatcher() PURE;
+
+  /**
+   * If an filter stopped filter iteration by returning FilterStatus::StopIteration,
```
"a filter"

jrajahalme replied:
Thanks.

mattklein123 commented on include/envoy/network/filter.h:
```diff
@@ -147,6 +148,67 @@ class FilterManager {
+  /**
+   * If an filter stopped filter iteration by returning FilterStatus::StopIteration,
+   * the filter should call continueFilterChain(true) when complete to continue the filter chain,
+   * or continueFilterChain(false) if the filter execution failed and the connection musrt be
```
typo "musrt"

jrajahalme replied:
Fixed.

mattklein123 commented on include/envoy/server/listener_manager.h:
```diff
@@ -169,7 +179,7 @@ class ListenerManager {
 
   /**
-   * Stop all listeners from accepting new connections without actually removing any of them. This
+   * Stop all listeners from accepting new connectins without actually removing any of them. This
```
typo/revert

jrajahalme replied:
Fixed.

mattklein123 commented on configs/original-dst-cluster/proxy_config.json:
```diff
@@ -1,7 +1,10 @@
     "listeners": [{
         "address": "tcp://0.0.0.0:10000",
-        "use_original_dst": true,
+        "listener_filters": [{
```
Can we please make this a v2 only feature? For v1 I would recommend just internally converting "use_original_dst" into a filter internally. Otherwise this is a breaking config change.

jrajahalme replied:
Happy to make this v2 only, simplifies it a little as well. It was not, however, a breaking config change, I had used this to test v1 access to the listener filters.

mattklein123 commented on source/common/filter/original_dst.cc:
```diff
@@ -0,0 +1,36 @@
+#include "common/filter/original_dst.h"
```
please put all listener filters in the `common/filter/listener` directory. It's bad that we don't already break things down like this. I would also use the `Envoy::Filter::Listener` namespace.

mattklein123 replied:
Also, it somehow possible to split this PR into the core code first, and then follow up PRs that independently convert proxy proto and original DST to filters? That might not be easily possible but I figured I would ask. :)

jrajahalme replied:
Will do the directory shuffle early next week. PR split is harder, as testing for correctness depends on the code being used. Also passing the coverage check might be problematic if the filters are split off.

mattklein123 commented on source/common/json/config_schemas.cc:
```diff
@@ -161,6 +161,15 @@ const std::string Json::Schema::LISTENER_SCHEMA(R"EOF(
         "required": ["cert_chain_file", "private_key_file"],
         "additionalProperties": false
       },
+      "listener_filters" : {
```
Please do not support this feature in v1.

jrajahalme replied:
Removed.

mattklein123 commented on include/envoy/network/listen_socket.h:
```diff
@@ -34,5 +34,65 @@ class ListenSocket {
+/**
+ * An abstract accept socket.
+ */
+class AcceptSocket {
```
I think I would call this `AcceptedSocket`. That's a little more natural for me. Please update related comments also.

jrajahalme replied:
Done.

mattklein123 commented on include/envoy/network/listen_socket.h:
```diff
@@ -34,5 +34,65 @@ class ListenSocket {
+   * Reset the original destination address of the socket to a different address than the one
+   * the socket was accepted at.
+   */
+  virtual void resetLocalAddress(Address::InstanceConstSharedPtr& local_address) PURE;
```
`const (Address::InstanceConstSharedPtr&`

jrajahalme replied:
Done

mattklein123 commented on include/envoy/network/listen_socket.h:
```diff
@@ -34,5 +34,65 @@ class ListenSocket {
+  /**
+   * Set the original source address of the socket
+   */
+  virtual void resetRemoteAddress(Address::InstanceConstSharedPtr& remote_address) PURE;
```
`const Address::InstanceConstSharedPtr&`

jrajahalme replied:
Ditto.

mattklein123 commented on include/envoy/network/filter.h:
```diff
@@ -147,6 +148,67 @@ class FilterManager {
+  virtual FilterStatus onAccept(ListenerFilterCallbacks& cb) PURE;
+};
+
+typedef std::shared_ptr<ListenerFilter> ListenerFilterSharedPtr;
```
I think this can probably be unique_ptr. The reason the other filters are shared_ptr is because of read/write/both semantics. That doesn't apply here.

jrajahalme replied:
Will try this out.

jrajahalme replied:
This was complicated a bit due to google mock not working with lvalue references. Worked around it by passing a raw pointer to addAcceptFilter() method.

mattklein123 commented on source/common/config/lds_json.cc:
```diff
@@ -57,6 +57,26 @@ void LdsJson::translateListener(const Json::Object& json_listener,
   JSON_UTIL_SET_BOOL(json_listener, *listener.mutable_deprecated_v1(), bind_to_port);
   JSON_UTIL_SET_INTEGER(json_listener, listener, per_connection_buffer_limit_bytes);
+
+  if (json_listener.hasObject("listener_filters")) {
```
please support v2 only.

mattklein123 commented on source/common/config/well_known_names.h:
```diff
@@ -47,6 +47,24 @@ class V1Converter {
+  const std::string PROXY_PROTOCOL = "envoy.proxy_protocol";
+
+  // Converts names from v1 to v2
+  const V1Converter v1_converter_;
```
v2 only please

jrajahalme replied:
OK.

mattklein123 commented on source/common/network/listen_socket_impl.h:
```diff
@@ -50,5 +50,48 @@ class UdsListenSocket : public ListenSocketImpl {
+    return fd;
+  }
+
+  void clearReset() override { local_address_reset_ = false; }
```
This existence/semantics of the `clearReset()` function is strange to me. It doesn't apply to remote, and it leaves the addresses to what they were reset to. This feels like a hack to me. Can you explain why this is needed and if there is a cleaner way to handle the need?

jrajahalme replied:
I was never happy with the name of this function either.

This is needed for handing off the accepted socket from one listener to another, which is needed for backwards compatibility in the case where the first listener listens on a redirected port and the second listener matches the original local address but does not bind.

Clearing of the local_address_reset_ bit is needed so that the second listener would not try to repeat the same check again (in an infinite recursion). See the call site in source/server/connection_handler_impl.cc.

This implementation also allows the second listener have and run it's own set of listener filters on the accepted socket it is given.

So the functionality is correct and needed. We should rename this function to lessen the likelihood of confusion in the future, though.

mattklein123 replied:
OK. I would probably switch reset with "change" and use the following function names (or some variant of them):
1) `changeLocalAddress(...)`
2) `localAddressChanged()`
3) `forgetOriginalLocalAddress()` (This should ASSERT that the local address was changed when called).

Etc.

And more heavily comment the interface. Feel free to iterate on my suggestions above, but that is more clear to me.

jrajahalme replied:
Just realized that the resetting the local_address_reset_ bit when passing the socket from one listener to another would break the original dst cluster from being able to forward the connection, so I'll have to figure alternative way of not recursing from the second listener. I don't see a need for using an original destination cluster in conjunction of non-bound listeners like this, but I don't like the way this is now anyway, so I'll spend some time on this. I'll try to come up with something that also addresses your concerns here and does not need heavy commenting :-)

mattklein123 commented on source/server/config/network/original_dst.cc:
```diff
@@ -0,0 +1,47 @@
+#include <string>
```
Please move the configs to a listener folder, like the filters.

jrajahalme replied:
Will do.

mattklein123 commented on source/server/connection_handler_impl.h:
```diff
@@ -47,35 +49,40 @@ class ConnectionHandlerImpl : public Network::ConnectionHandler, NonCopyable {
     ~ActiveListener();
 
+    // Network::ListenerCallbacks
+    /**
```
nit: don't need to duplicate doc comments here

mattklein123 commented on source/server/listener_manager_impl.cc:
```diff
@@ -48,6 +48,45 @@ ProdListenerComponentFactory::createFilterFactoryList_(
+            string_name);
+    if (factory != nullptr) {
+      Configuration::ListenerFilterFactoryCb callback;
+      if (filter_config->getBoolean("deprecated_v1", false)) {
```
v2 only please

jrajahalme replied:
Done.

ggreenway commented:
I looked through this, and I wonder if it should be reworked so that there is a connection object of some type (a thing to hold the fd, and handle reading from it) is created as soon as the connection is accepted.  It would then call the new set of filters, then match which logical listener and/or filter chain to use, then proceed.

The piece about this I dislike, which made me think of the above, is that the proxy protocol filter has to implement reading from the socket directly.  I imagine other listener filters will need to do the same.  

The other thing I think we'll want eventually, is for one of these new filters to read some data, look at it, and then pass it along (modified or not) to the network filters.  This could be done with MSG_PEEK, but that seems like a hack.

Thoughts?

lizan commented on include/envoy/server/filter_config.h:
```diff
@@ -123,6 +123,62 @@ class FactoryContext {
+   *         JSON and then parsed into this empty proto. Optional today, will be compulsory when v1
+   *         is deprecated.
+   */
+  virtual ProtobufTypes::MessagePtr createEmptyConfigProto() { return nullptr; }
```
this should be a PURE method.

jrajahalme replied:
Right, fixed!

lizan commented on source/server/config/network/original_dst.cc:
```diff
@@ -0,0 +1,47 @@
+  }
+
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override {
+    return ProtobufTypes::MessagePtr{new Envoy::ProtobufWkt::Empty()};
```
nit: `std::make_unique`

jrajahalme replied:
Envoy::ProtobufWkt::Empty() is a more specific type than ProtobufTypes::Message, do you think using std::make_unique would make this easier on the eyes?

lizan commented on source/server/connection_handler_impl.cc:
```diff
@@ -5,29 +5,20 @@
-  ActiveListenerPtr l(new SslActiveListener(*this, ssl_ctx, socket, factory, scope, listener_tag,
-                                            listener_options));
-  listeners_.emplace_back(socket.localAddress(), std::move(l));
+void ConnectionHandlerImpl::addListener(Listener& config) {
```
Thanks for consolidating addListner and addSslListener, neat!

lizan commented on source/server/connection_handler_impl.h:
```diff
@@ -130,6 +137,36 @@ class ConnectionHandlerImpl : public Network::ConnectionHandler, NonCopyable {
+  /**
+   * Wrapper for an active accept socket owned by this handler.
+   */
+  struct ActiveSocket : public Network::ListenerFilterManager,
```
nit: this still have to be a class, per Google Style Guide (no deviation in envoy style for this):

`Use a struct only for passive objects that carry data; everything else is a class.`

jrajahalme replied:
I tried this but this happened even when I made all members public:
```
source/server/connection_handler_impl.cc: In destructor 'virtual Envoy::Server::ConnectionHandlerImpl::ActiveListener::~ActiveListener()':
source/server/connection_handler_impl.cc:72:72: error: 'std::unique_ptr<_Tp> Envoy::LinkedObject<T>::removeFromList(Envoy::LinkedObject<T>::ListType&) [with T = Envoy::Server::ConnectionHandlerImpl::ActiveSocket; Envoy::LinkedObject<T>::ListType = std::__cxx11::list<std::unique_ptr<Envoy::Server::ConnectionHandlerImpl::ActiveSocket> >]' is inaccessible within this context
     ActiveSocketPtr removed = sockets_.front()->removeFromList(sockets_);
```

Maybe there is a reason other structs like this also exist in this file?

lizan commented on source/server/listener_manager_impl.cc:
```diff
@@ -48,6 +48,45 @@ ProdListenerComponentFactory::createFilterFactoryList_(
+    Configuration::FactoryContext& context) {
+  std::vector<Configuration::ListenerFilterFactoryCb> ret;
+  for (ssize_t i = 0; i < filters.size(); i++) {
+    const std::string string_type = filters[i].deprecated_v1().type();
```
nit: `ProtobufTypes::String`

jrajahalme replied:
Fixed.

lizan commented on source/server/listener_manager_impl.h:
```diff
@@ -30,13 +30,25 @@ class ProdListenerComponentFactory : public ListenerComponentFactory,
-  // Server::ListenSocketFactory
+  // Server::ListenerComponentFactory
   std::vector<Configuration::NetworkFilterFactoryCb>
   createFilterFactoryList(const Protobuf::RepeatedPtrField<envoy::api::v2::Filter>& filters,
```
rename this to `createNetworkFilterFactoryList`?

lizan commented on include/envoy/network/filter.h:
```diff
@@ -161,6 +223,14 @@ class FilterChainFactory {
+   * @return true if filter chain was created successfully. Otherwise
+   *   false.
+   */
+  virtual bool createFilterChain(ListenerFilterManager& listener) PURE;
```
`createListnerFilterChain` for this and `createNetworkFilterChain` for the one above, this shouldn't be an overloaded.

jrajahalme replied:
OK.

jrajahalme commented on include/envoy/server/filter_config.h:
```diff
@@ -123,6 +123,62 @@ class FactoryContext {
+   * @param context supplies the filter's context.
+   * @return ListenerFilterFactoryCb the factory creation function.
+   */
+  virtual ListenerFilterFactoryCb createFilterFactory(const Json::Object& config,
```
OK, removed.

mattklein123 commented on include/envoy/network/listen_socket.h:
```diff
@@ -34,5 +34,65 @@ class ListenSocket {
+};
+
+typedef std::unique_ptr<AcceptedSocket> AcceptedSocketPtr;
+typedef std::shared_ptr<AcceptedSocket> AcceptedSocketSharedPtr;
```
Is this shared_ptr definition needed?

jrajahalme replied:
No :-)

mattklein123 commented on include/envoy/network/listen_socket.h:
```diff
@@ -34,5 +34,65 @@ class ListenSocket {
+  /**
+   * @return fd the accepted socket's file descriptor.
+   */
+  virtual int fd() const PURE;
```
Per @ggreenway, can we hide the fd from the filters and potentially just expose peek operations for now? I feel like fd can be an implementation detail and passed around as needed into the real connection when it is created. In general, we should heavily limit what filters can do on a non-const accepted socket. Can we try to limit this interface as much as possible, at least from the filter perspective? This might require a base interface just for use by filters.

PiotrSikora replied:
Note that PROXY protocol consumes bytes from the wire, so peek operation alone wouldn't work (but it's also needed).

mattklein123 replied:
Right, good point. We can have a read, etc. method also if needed.

jrajahalme replied:
Socket option access is needed by the original dst filter, so that should be added in as well. On the other hand I don't know what the transport socket PR will bring. Maybe keep these listener filters and the interface as they are now and reconsider after the transport socket work is in?

PiotrSikora replied:
I agree that leaving `fd()` in this PR should be fine (unless @mattklein123 and/or @ggreenway have strong feelings against it), just please add `TODO` and hide the socket-level access in another PR (perhaps even after all other changes are merged?).

mattklein123 replied:
I think it's fine to leave `fd()` in this PR, but per @PiotrSikora please add a TODO. We should also remove the fd transfer/move function from the public interface. That should be an internal implementation detail. Then in a followup we can remove `fd()` and replace with the few accessor functions we need.

jrajahalme replied:
Added a TODO.

mattklein123 commented on include/envoy/network/filter.h:
```diff
@@ -147,6 +148,67 @@ class FilterManager {
+   * first is called first).
+   * @param filter supplies the filter being added.
+   */
+  virtual void addAcceptFilter(ListenerFilter* filter) PURE;
```
This should take a `ListenerFilterPtr&&`. If that is not possible for some reason, we should use shared_ptr.

jrajahalme replied:
The trouble is with gmock not being able to mock functions with lvalue references. Apparently that is something they have decided they will never support.

mattklein123 replied:
We handle this in many different places. There is a common pattern on how to do this. See https://github.com/envoyproxy/envoy/blob/master/test/mocks/event/mocks.h#L67 for example.

jrajahalme replied:
Fixed, thanks for the pointer!

mattklein123 commented on include/envoy/network/filter.h:
```diff
@@ -160,7 +222,15 @@ class FilterChainFactory {
+   * Called to create the listener filter chain.
+   * @param listener supplies the listener to create the chain on.
+   * @return true if filter chain was created successfully. Otherwise
+   *   false.
```
nit: reflow comment

jrajahalme replied:
Not exactly sure what you mean here, but I moved the last word to the end of the previous line.

mattklein123 commented on include/envoy/network/listen_socket.h:
```diff
@@ -34,5 +34,65 @@ class ListenSocket {
+  /**
+   * Close the underlying socket.
+   */
+  virtual void close() PURE;
```
Do we need an explicit `close()` method? Can this be handled in destruction? Even in the case of deferred deletion it shouldn't matter.

jrajahalme replied:
Right, I'll remove the explicit close().

jrajahalme replied:
Had to reintroduce this in order to remove takeFd(), which was more offensive.

mattklein123 commented on source/server/connection_handler_impl.cc:
```diff
@@ -63,71 +54,140 @@ void ConnectionHandlerImpl::ActiveListener::removeConnection(ActiveConnection& c
+          // Reset the accepted socket transient state and hand it to the new listener.
+          socket_->clearReset();
+          new_listener->onAccept(std::move(socket_));
+          goto out;
```
Please try to restructure the code to avoid goto.

jrajahalme replied:
I'm working on this, will update the PR today.

mattklein123 commented on source/server/connection_handler_impl.h:
```diff
@@ -47,37 +49,43 @@ class ConnectionHandlerImpl : public Network::ConnectionHandler, NonCopyable {
+    // Network::ListenerCallbacks
     /**
-     * Fires when a new connection is received from the listener.
+     * Fires when a new accepted socket is received from the listener.
```
nit: delete doc comments for overridden functions

jrajahalme replied:
OK.

rshriram commented:
This is nice.. Just one thought: why can't we move the filter chain match as a "filter" into the listener filter framework? Its after all enhancing the stock listener by selecting the L7 filter chain (tcp/http/etc.)..

Second question: is there a way to short circuit the original_dst_cluster style code at the listener level, so that we don't even have to enter the tcp/http conn man ? This is typically useful in TCP proxy scenarios, where you trap all traffic and send to Envoy. You install exceptions in the listener to discard/passthrough traffic for certain IP/port pairs. I don't know if there is a way to signal to the kernel to not send packets for this connection back into Envoy again. But if we could, it would essentially give us a poor man's OVS.

jrajahalme commented on include/envoy/network/listen_socket.h:
```diff
@@ -34,5 +34,65 @@ class ListenSocket {
+  virtual Address::InstanceConstSharedPtr localAddress() const PURE;
+
+  /**
+   * Reset the original destination address of the socket to a different address than the one
```
OK, doing the same for resetRemoteAddress(), too.

jrajahalme commented:
> On Jan 15, 2018, at 10:50 AM, Shriram Rajagopalan <notifications@github.com> wrote:
> 
> This is nice.. Just one thought: why can't we move the filter chain match as a "filter" into the listener filter framework? Its after all enhancing the stock listener by selecting the L7 filter chain (tcp/http/etc.)..
> 
> 

Haven’t looked into this but doing this may require exposing too much of the listener internals to the filters.
> Second question: is there a way to short circuit the original_dst_cluster style code at the listener level, so that we don't even have to enter the tcp/http conn man ? This is typically useful in TCP proxy scenarios, where you trap all traffic and send to Envoy. You install exceptions in the listener to discard/passthrough traffic for certain IP/port pairs. I don't know if there is a way to signal to the kernel to not send packets for this connection back into Envoy again. But if we could, it would essentially give us a poor man's OVS.
> 
> 

IMO you would need a kernel side TCP proxy that plays nice with Envoy to even try this :-)

  Jarno

jrajahalme commented:
Just posted a prerequisite PR #2382

mattklein123 commented on include/envoy/network/listen_socket.h:
```diff
@@ -34,5 +34,55 @@ class ListenSocket {
+   * Transfer ownership of the file descriptor to the caller, so that the underlying socket will
+   * not be closed on delete.
+   */
+  virtual int takeFd() PURE;
```
Per my other comment let's remove this from the exposed interface. I think we should be able to hide this as an implementation detail during listener processing (one option is to just store the `AcceptedSocket and actually pass it, vs. the `fd` to connections). If this is not easily possible can we add a TODO to think about cleaning this up? At the very least I don't want to pass this to filters, so we might need to have a split interface for internal/listener use and for filter use.

jrajahalme replied:
OK, did this, but it was a lot of work and I had to bring close() back in to the exposed interface. Had to split up AcceptedSocket into multiple classes as it felt wrong to use that for ClientConnection. Let me know if you'd rather revert the last commit and do it as a separate PR after this one.

mattklein123 commented:
@jrajahalme I think we are pretty close on the interface. Do you mind merging master and following up on the remaining interface comments? Then I think we can begin digging into the implementation details.

jrajahalme commented:
> On Jan 16, 2018, at 2:15 PM, Matt Klein <notifications@github.com> wrote:
> 
> @jrajahalme <https://github.com/jrajahalme> I think we are pretty close on the interface. Do you mind merging master and following up on the remaining interface comments? Then I think we can begin digging into the implementation details.
> 

I think I have addressed all comments so far now :-)

mattklein123 commented on source/common/json/config_schemas.cc:
```diff
@@ -178,6 +178,10 @@ const std::string Json::Schema::LISTENER_SCHEMA(R"EOF(
     "properties": {
        "name": {"type": "string"},
        "address": {"type": "string"},
+       "listener_filters" : {
```
Can be reverted.

jrajahalme replied:
Done, thanks for pointing this out.

ggreenway commented on include/envoy/network/listen_socket.h:
```diff
@@ -34,5 +34,75 @@ class ListenSocket {
+
+typedef std::unique_ptr<AcceptedSocket> AcceptedSocketPtr;
+
+class ClientSocket : virtual public ConnectionSocket {
```
Can you please add some comments on what this interface is for?

jrajahalme commented:
> On Jan 17, 2018, at 9:57 AM, Greg Greenway <notifications@github.com> wrote:
> 
> @ggreenway commented on this pull request.
> 
> I don't have time to look carefully right now, but I took a brief look through the interfaces and I don't have any major objections; overall looks cleaner than it was before.
> 
> In include/envoy/network/listen_socket.h <https://github.com/envoyproxy/envoy/pull/2346#discussion_r162128999>:
> 
> > +  /**
> +   * Reset the destination address of the socket to a different address than the one
> +   * the socket was accepted at.
> +   */
> +  virtual void resetLocalAddress(const Address::InstanceConstSharedPtr& local_address) PURE;
> +
> +  /**
> +   * Reset the source address of the socket to a different address than the one
> +   * the socket was accepted at.
> +   */
> +  virtual void resetRemoteAddress(const Address::InstanceConstSharedPtr& remote_address) PURE;
> +};
> +
> +typedef std::unique_ptr<AcceptedSocket> AcceptedSocketPtr;
> +
> +class ClientSocket : virtual public ConnectionSocket {
> Can you please add some comments on what this interface is for?
> 
> 

Added.

mattklein123 commented on source/common/network/connection_impl.cc:
```diff
@@ -540,9 +478,48 @@ ClientConnectionImpl::ClientConnectionImpl(
+  // The local address can only be retrieved for IP connections. Other
+  // types, such as UDS, don't have a notion of a local address.
+  if (socket_->remoteAddress()->type() == Address::Type::Ip) {
+    dynamic_cast<ClientSocket*>(socket_.get())->setLocalAddress(Address::addressFromFd(fd()));
```
not super crazy about this dynamic cast. I honestly don't recall at this point why I wrote this code the way I did (with connect/bind in the base class vs. client class). Probably to deal with some of the issues that you are dealing with here. What was the rationale for splitting it out? I agree it's better to have this be part of the client class, but then we face some of these cast issues. Just curious on your thought process here (I'm not sure if right solution). Instead of dynamic cast should we just allow changing local address on connection socket?

jrajahalme replied:
Splitting the functionality was not the reason for the dynamic cast, but the hierarchy of the new Socket classes was. I refactored the abstract Socket classes back to one (ConnectionSocket) removing the need for the dynamic cast. There was no need for the hierarchy now that setLocalAddress and resetLocalAddress are combined to the same method.

mattklein123 commented on source/common/network/listen_socket_impl.h:
```diff
@@ -50,5 +50,63 @@ class UdsListenSocket : public ListenSocketImpl {
+  AcceptedSocketImpl(int fd, const Address::InstanceConstSharedPtr& local_address,
+                     const Address::InstanceConstSharedPtr& remote_address)
+      : ConnectionSocketImpl(fd, local_address, remote_address) {}
+  ~AcceptedSocketImpl() {}
```
nit: del empty destructor

jrajahalme replied:
Done.

mattklein123 commented on source/common/network/listen_socket_impl.h:
```diff
@@ -50,5 +50,63 @@ class UdsListenSocket : public ListenSocketImpl {
+  ~ClientSocketImpl() {}
+
+  // ClientSocket
+  void setLocalAddress(const Address::InstanceConstSharedPtr& local_address) override {
```
The existence of `setLocalAddress` and `resetLocalAddress` seems pretty strange to me. Can we just have `changeLocalAddress` in the base class? Would this solve some of the other problems I've commented on?

mattklein123 commented on source/common/network/listen_socket_impl.h:
```diff
@@ -50,5 +50,63 @@ class UdsListenSocket : public ListenSocketImpl {
+
+  // Network::AcceptedSocket
+  void resetLocalAddress(const Address::InstanceConstSharedPtr& local_address) override {
+    if (*local_address != *local_address_) {
```
nit: The comparison here is almost definitely slower than just assigning universally. Or is this to prevent reset from being set? If so comment on how/why/when this happens?

jrajahalme replied:
It was to prevent the flag being set, but I refactored this and made it the callers responsibility to make sure the addresses are actually different when that matters.

mattklein123 commented on source/common/network/listen_socket_impl.h:
```diff
@@ -50,5 +50,63 @@ class UdsListenSocket : public ListenSocketImpl {
+  ClientSocketImpl(const Address::InstanceConstSharedPtr& remote_address)
+      : ConnectionSocketImpl(remote_address->socket(Address::SocketType::Stream), nullptr,
+                             remote_address) {}
+  ~ClientSocketImpl() {}
```
nit: del empty constructor

mattklein123 commented on source/common/network/listener_impl.cc:
```diff
@@ -23,77 +20,39 @@ Address::InstanceConstSharedPtr ListenerImpl::getLocalAddress(int fd) {
+      fd,
+      // Get the local address from the new socket if the listener is listening on the all hosts
+      // address (e.g., 0.0.0.0 for IPv4).
+      !listener->local_address_ ? listener->getLocalAddress(fd) : listener->local_address_,
```
So local_address_ is nullptr for 0.0.0.0? Or it's the ANY address? I thought it was the latter? Maybe I just don't remember. Can you add to comment that ANY address is nullptr if that is the case? OK I see below you changed this. Can you just add comment here that nullptr indicates any address?

mattklein123 commented on source/common/network/listener_impl.cc:
```diff
@@ -23,77 +20,39 @@ Address::InstanceConstSharedPtr ListenerImpl::getLocalAddress(int fd) {
+ListenerImpl::ListenerImpl(Event::DispatcherImpl& dispatcher, ListenSocket& socket,
+                           ListenerCallbacks& cb, bool bind_to_port)
+    : local_address_(nullptr), cb_(cb), listener_(nullptr) {
+  auto ip = socket.localAddress()->ip();
```
nit: const

mattklein123 commented on source/server/config/listener/original_dst.cc:
```diff
@@ -0,0 +1,42 @@
+                                                       FactoryContext&) override {
+    return [](Network::ListenerFilterManager& filter_manager) -> void {
+      filter_manager.addAcceptFilter(
+          Network::ListenerFilterPtr{new Filter::Listener::OriginalDst()});
```
nit: std::make_unique (same elsewhere)

mattklein123 commented on source/server/connection_handler_impl.h:
```diff
@@ -88,19 +86,19 @@ class ConnectionHandlerImpl : public Network::ConnectionHandler, NonCopyable {
-                      Stats::Scope& scope, uint64_t listener_tag,
-                      const Network::ListenerOptions& listener_options);
+    Network::ListenerConfig& config_;
+    Filter::Listener::ProxyProtocol::ConfigSharedPtr legacy_stats_;
```
Instead of doing it this way, can we do legacy by actually altering the configuration to create filters configs? That would unify all of the logic and be cleaner IMO.

jrajahalme replied:
Will do this tomorrow.

jrajahalme replied:
Should we remove the legacy flag support ('use_original_dst', 'use_proxy_protocol') from the v2 API, so do you want to keep them for convenience?

mattklein123 replied:
I would probably do one of two things:
1) Don't document original DST / proxy proto as filters, but just instantiate them as such. (So no change to documentation for those filters).
2) Deprecate the options and then we can delete support for them in the 1.7.0 cycle.

I don't really have a strong opinion. @htuch or @alyssawilk any opinion? (The question is how to deal with original DST and proxy proto legacy config).

jrajahalme replied:
One complication here is that in v2 lds config 'use_proxy_proto' flag is in the filter_chain rather than in the config root. I guess that this is done this way to make it possible to filter on the source address against a known ELB address and then select a filter_chain with proxy protocol, which would allow ELB and non-ELB traffic to be accepted by the same listener on the same port.

Filter chain matching is not yet implemented, so this issue is not apparent in the code as of now, but my design in this PR runs both proxy protocol and original dst filters in the same listener_filters list, which would be run to completion before the filter_chain matching is (to be) done. This is why the listener_filters are specified as part of the main listener config rather than a filter_chain config.

The reasoning for having the listener filters before the filter_chain matching is that in the case of use_original_dst, it is more meaningful to match on the destination address for filter_chain selection after the destination address has been restored. With regard to the destination address this same argument holds for proxy protocol, too. I had not thought about the source address filtering for proxy protocol before, but it seems it would be beneficial to be able to do filter chain matching both before and after running proxy protocol, first on the source address to figure out that you need to run proxy protocol, and then after to select the network filter chain based on the restored destination address.

I guess the most flexible work around for this would be to allow insertion of listener filters both at the listener and filter_chain level. Thoughts?

mattklein123 replied:
@jrajahalme ACK, this seems more complicated than I thought. I'm investigating and will get back to you with some thoughts.

mattklein123 replied:
Ugh. This is a mess. Thanks for pointing this out @jrajahalme. Honestly, I don't feel it's worth the extra complexity to support running the listener filters more than once. To me it's an edge case, unless other people feel really strongly about it.

If we want to support proxy proto at the filter chain level, I wonder if it could be a special case (not an exposed filter) that runs after the filter chain is matched? Otherwise, I'm kind of the opinion that we should move proxy_proto into the base listener and deprecate where it is now.

@htuch @ggreenway @lizan @PiotrSikora thoughts ^^^ ?

mattklein123 replied:
> Deprecating both bool flags could take care of this problem as well? This is not an issue for v1, and it is not an issue for anyone using v2 API right now, as the filter chain matching is not implemented yet. I mean that maybe even that special casing for proxy_proto after filter chain match is not needed?

Yes, I would argue it is not explicitly needed (I can see why someone would ask for it, but I am skeptical of how often this would come about). I would be fine with deprecating both booleans and moving this into the base listener level only.

My concern really is that someone feels very strongly that we need to support the use case where we do proxy_proto depending on filter chain matching, so looking for feedback from the folks that I already mentioned.

mattklein123 replied:
@jrajahalme I discussed with @PiotrSikora @lizan @ggreenway and we think that we should leave proxy proto on the filter chain, but we don't want to do extra matching after we run it. So basically, proxy proto should not be a normal filter that is documented and run as part of the listener filter chain.

I think internally, it's probably OK if this is a filter, we can use it as such and special case it later when we implement real filter chain matching. Here is what I would do concretely:

1) Remove proxy proto filter from docs. The fact that it is a filter is an implementation detail.
2) For now just install it as the last filter as you are doing now.
3) I would do the configuration conversion as we previously discussed, assuming a single filter chain match and just put in TODOs for fix it up later.

Does that make sense?

mattklein123 commented on source/server/connection_handler_impl.cc:
```diff
@@ -69,71 +60,141 @@ void ConnectionHandlerImpl::ActiveListener::removeConnection(ActiveConnection& c
+void ConnectionHandlerImpl::ActiveListener::onAccept(Network::AcceptedSocketPtr&& socket,
+                                                     bool redirected) {
+  Network::Address::InstanceConstSharedPtr local_address = socket->localAddress();
+  ActiveSocket* active_socket(new ActiveSocket(*this, std::move(socket), redirected));
```
std::unique_ptr (please no naked pointers anywhere)

jrajahalme replied:
Fixed by linking the unique_ptr only if filter iteration does not conclude on the first call to continueFIlterChain().

mattklein123 commented on source/server/connection_handler_impl.cc:
```diff
@@ -69,71 +60,141 @@ void ConnectionHandlerImpl::ActiveListener::removeConnection(ActiveConnection& c
+      legacy_stats_(new Filter::Listener::ProxyProtocol::Config(config.listenerScope())) {}
 
 ConnectionHandlerImpl::ActiveListener::~ActiveListener() {
+  while (!sockets_.empty()) {
```
these are sockets that are still in filter iteration but have not been converted to connections, right? Can you add some comments?

jrajahalme replied:
Right, added a comment.

mattklein123 commented on source/server/connection_handler_impl.cc:
```diff
@@ -69,71 +60,141 @@ void ConnectionHandlerImpl::ActiveListener::removeConnection(ActiveConnection& c
   // However, linear performance might be adequate since the number of listeners is small.
   // We do not return stopped listeners.
-  auto listener = std::find_if(
+  auto iter = std::find_if(
```
nit: IMO `listener` is more clear. if you want to change I would do `listener_it`

jrajahalme replied:
I found 'listener' confusing. Changed to 'listener_it'

mattklein123 commented on source/server/connection_handler_impl.cc:
```diff
@@ -69,71 +60,141 @@ void ConnectionHandlerImpl::ActiveListener::removeConnection(ActiveConnection& c
+
+  // Filter execution concluded, clear state.
+  iter_ = accept_filters_.end();
+  ActiveSocketPtr removed = removeFromList(listener_->sockets_);
```
this is always nullptr? Right? (Since we moved it). Can we assert that and drop the deferred delete call?

jrajahalme replied:
No, we moved the AcceptedSocket to the connection or another listener, this is about cleaning up the ActiveSocket itself. I had to change this a bit due to the bare pointer fix below, however.

mattklein123 commented on source/server/connection_handler_impl.h:
```diff
@@ -130,6 +128,38 @@ class ConnectionHandlerImpl : public Network::ConnectionHandler, NonCopyable {
+    Event::Dispatcher& dispatcher() override { return listener_->parent_.dispatcher_; }
+    void continueFilterChain(bool success) override;
+
+    ActiveListener* listener_;
```
nit: `ActiveListener&`

jrajahalme replied:
Fixed. Have to get to the habit of avoiding bare pointers...

mattklein123 commented on source/server/http/admin.h:
```diff
@@ -162,6 +165,30 @@ class AdminImpl : public Admin,
+    const std::string& name() const override { return name_; }
+
+    AdminImpl& parent_;
+    std::string name_;
```
const

jrajahalme replied:
Fixed.

PiotrSikora commented on include/envoy/network/listen_socket.h:
```diff
@@ -34,5 +34,75 @@ class ListenSocket {
+  virtual const Address::InstanceConstSharedPtr& remoteAddress() const PURE;
+
+  /**
+   * @return true if the local address has been reset.
```
+1

jrajahalme replied:
I must have missed that suggestion. The real semantics is that the original destination cluster needs to know if the local address was "restored to it's original value, as it was before redirection". The significance of this is that missing that distinction will cause the upstream connection be opened to the same local address the Listener was listening at leading to recursion of open connections until Envoy reaches the max number of open connections and dies. This also explains the difference between the resetLocalAddress() and setLocalAddress(). The latter is used for client connections, where the local address is fully known only after the socket is connected.

So, initially the local address is the one the listener received the new connection at, but it can be restored to the original destination address by the original dst listener filter.

Maybe it would be clearer to add a separate flag to the address change method to mark as the address as 'restored'? This way the semantics would be more explicit and likely easier to understand. I'll try this out.

jrajahalme commented on include/envoy/network/filter.h:
```diff
@@ -147,6 +148,67 @@ class FilterManager {
+  virtual ~ListenerFilterCallbacks() {}
+
+  /**
+   * @return the AcceptedSocket that owns this manager and the managed filters.
```
Reworded as "@return AcceptedSocket the socket the filter is operating on."

jrajahalme commented on include/envoy/network/filter.h:
```diff
@@ -147,6 +148,67 @@ class FilterManager {
+  virtual AcceptedSocket& socket() PURE;
+
+  /**
+   * @return the Dispatcher for this manager.
```
This better: "@return the Dispatcher for issuing events."?

jrajahalme commented on include/envoy/network/listen_socket.h:
```diff
@@ -34,5 +34,75 @@ class ListenSocket {
+  virtual ~AcceptedSocket() {}
+
+  /**
+   * Reset the destination address of the socket to a different address than the one
```
I'm trying out "set"...

jrajahalme commented on source/common/config/well_known_names.h:
```diff
@@ -47,6 +47,19 @@ class V1Converter {
 };
 
+/**
+ * Well-known network filter names.
```
Oops, fixed.

jrajahalme commented on source/common/config/well_known_names.h:
```diff
@@ -47,6 +47,19 @@ class V1Converter {
+class ListenerFilterNameValues {
+public:
+  // Original destination listener filter
+  const std::string ORIGINAL_DST = "envoy.original_dst";
```
Agree.

jrajahalme commented on source/common/filter/listener/original_dst.cc:
```diff
@@ -0,0 +1,38 @@
+Network::FilterStatus OriginalDst::onAccept(Network::ListenerFilterCallbacks& cb) {
+  ENVOY_LOG(debug, "original_dst: New connection accepted");
+  Network::AcceptedSocket& socket = cb.socket();
+  Network::Address::InstanceConstSharedPtr local_address = socket.localAddress();
```
Good point, done!

jrajahalme commented on source/common/filter/listener/original_dst.h:
```diff
@@ -0,0 +1,24 @@
+  // Network::ListenerFilter
+  Network::FilterStatus onAccept(Network::ListenerFilterCallbacks& cb) override;
+
+  virtual Network::Address::InstanceConstSharedPtr getOriginalDst(int fd);
```
Done.

jrajahalme commented on source/common/filter/listener/proxy_protocol.h:
```diff
@@ -0,0 +1,83 @@
+ */
+class Instance : public Network::ListenerFilter, Logger::Loggable<Logger::Id::filter> {
+public:
+  Instance(ConfigSharedPtr config) : search_index_(1), config_(config) {}
```
Done.

jrajahalme commented:
> On Jan 18, 2018, at 12:05 PM, Matt Klein <notifications@github.com> wrote:
> 
> @mattklein123 commented on this pull request.
> 
> In source/server/connection_handler_impl.h <https://github.com/envoyproxy/envoy/pull/2346#discussion_r162454438>:
> 
> >      std::list<ActiveConnectionPtr> connections_;
>      const uint64_t listener_tag_;
> -  };
> -
> -  struct SslActiveListener : public ActiveListener {
> -    SslActiveListener(ConnectionHandlerImpl& parent, Ssl::ServerContext& ssl_ctx,
> -                      Network::ListenSocket& socket, Network::FilterChainFactory& factory,
> -                      Stats::Scope& scope, uint64_t listener_tag,
> -                      const Network::ListenerOptions& listener_options);
> +    Network::ListenerConfig& config_;
> +    Filter::Listener::ProxyProtocol::ConfigSharedPtr legacy_stats_;
> Ugh. This is a mess. Thanks for pointing this out @jrajahalme <https://github.com/jrajahalme>. Honestly, I don't feel it's worth the extra complexity to support running the listener filters more than once. To me it's an edge case, unless other people feel really strongly about it.
> 
> If we want to support proxy proto at the filter chain level, I wonder if it could be a special case (not an exposed filter) that runs after the filter chain is matched? Otherwise, I'm kind of the opinion that we should move proxy_proto into the base listener and deprecate where it is now.
> 
> @htuch <https://github.com/htuch> @ggreenway <https://github.com/ggreenway> @lizan <https://github.com/lizan> @PiotrSikora <https://github.com/piotrsikora> thoughts ^^^ ?
> 
> 

Deprecating both bool flags could take care of this problem as well? This is not an issue for v1, and it is not an issue for anyone using v2 API right now, as the filter chain matching is not implemented yet. I mean that maybe even that special casing for proxy_proto after filter chain match is not needed?

jrajahalme commented:
> On Jan 18, 2018, at 2:38 PM, Matt Klein <notifications@github.com> wrote:
> 
> @mattklein123 commented on this pull request.
> 
> In source/server/connection_handler_impl.h <https://github.com/envoyproxy/envoy/pull/2346#discussion_r162491253>:
> 
> >      std::list<ActiveConnectionPtr> connections_;
>      const uint64_t listener_tag_;
> -  };
> -
> -  struct SslActiveListener : public ActiveListener {
> -    SslActiveListener(ConnectionHandlerImpl& parent, Ssl::ServerContext& ssl_ctx,
> -                      Network::ListenSocket& socket, Network::FilterChainFactory& factory,
> -                      Stats::Scope& scope, uint64_t listener_tag,
> -                      const Network::ListenerOptions& listener_options);
> +    Network::ListenerConfig& config_;
> +    Filter::Listener::ProxyProtocol::ConfigSharedPtr legacy_stats_;
> @jrajahalme <https://github.com/jrajahalme> I discussed with @PiotrSikora <https://github.com/piotrsikora> @lizan <https://github.com/lizan> @ggreenway <https://github.com/ggreenway> and we think that we should leave proxy proto on the filter chain, but we don't want to do extra matching after we run it. So basically, proxy proto should not be a normal filter that is documented and run as part of the listener filter chain.
> 
> I think internally, it's probably OK if this is a filter, we can use it as such and special case it later when we implement real filter chain matching. Here is what I would do concretely:
> 
> Remove proxy proto filter from docs. The fact that it is a filter is an implementation detail.
> For now just install it as the last filter as you are doing now.
> I would do the configuration conversion as we previously discussed, assuming a single filter chain match and just put in TODOs for fix it up later.
> Does that make sense?
> 
> 

Yes. Will post updates soon.

jrajahalme commented:
I don't have an OSX build setup locally, so bear with me for a few rounds of debugging here. Alternatively I would appreciate if someone pointed me towards instructions on how to run a single test locally on a Mac.

mattklein123 commented:
@jrajahalme @zuercher might be able to help out. You might want to grab him on Slack.

lizan commented:
I can help too, ping me on Slack.

jrajahalme commented:
There must be some race condition in OSX socket/connection IO. On an earlier run the "Basic/0" (IPv4) hang, now it passed, but "Basic/1" (IPv6) hangs:
```
==================== Test output for //test/common/network:proxy_protocol_test:
[==========] Running 42 tests from 2 test cases.
[----------] Global test environment set-up.
[----------] 38 tests from IpVersions/ProxyProtocolTest
[ RUN      ] IpVersions/ProxyProtocolTest.Basic/0
[2018-01-19 22:20:45.769][174619][debug][connection] source/common/network/connection_impl.cc:499] [C1] connecting to 127.0.0.1:53397
[2018-01-19 22:20:45.770][174619][debug][connection] source/common/network/connection_impl.cc:508] [C1] connection in progress
[2018-01-19 22:20:45.771][174619][debug][filter] source/common/filter/listener/proxy_protocol.cc:28] proxy_protocol: New connection accepted
[2018-01-19 22:20:45.771][174619][trace][connection] source/common/network/connection_impl.cc:342] [C1] socket event: 2
[2018-01-19 22:20:45.771][174619][trace][connection] source/common/network/connection_impl.cc:403] [C1] write ready
[2018-01-19 22:20:45.771][174619][debug][connection] source/common/network/connection_impl.cc:413] [C1] connected
[2018-01-19 22:20:45.772][174619][trace][connection] source/common/network/connection_impl.cc:279] [C1] writing 56 bytes
[2018-01-19 22:20:45.772][174619][trace][connection] source/common/network/connection_impl.cc:342] [C1] socket event: 2
[2018-01-19 22:20:45.772][174619][trace][connection] source/common/network/connection_impl.cc:403] [C1] write ready
[2018-01-19 22:20:45.772][174619][trace][connection] source/common/network/raw_buffer_socket.cc:58] [C1] write returns: 56
[2018-01-19 22:20:45.772][174619][debug][filter] source/common/filter/listener/proxy_protocol.cc:152] proxy_protocol: Read 47 bytes (total 47) from socket (found: true)
[2018-01-19 22:20:45.773][174619][debug][main] source/server/connection_handler_impl.cc:193] [C2] new connection
[2018-01-19 22:20:45.773][174619][trace][main] source/common/event/dispatcher_impl.cc:126] item added to deferred deletion list (size=1)
[2018-01-19 22:20:45.773][174619][trace][main] source/common/event/dispatcher_impl.cc:53] clearing deferred deletion list (size=1)
[2018-01-19 22:20:45.773][174619][trace][connection] source/common/network/connection_impl.cc:342] [C1] socket event: 2
[2018-01-19 22:20:45.773][174619][trace][connection] source/common/network/connection_impl.cc:403] [C1] write ready
[2018-01-19 22:20:45.773][174619][trace][connection] source/common/network/connection_impl.cc:342] [C2] socket event: 2
[2018-01-19 22:20:45.773][174619][trace][connection] source/common/network/connection_impl.cc:403] [C2] write ready
[2018-01-19 22:20:45.773][174619][debug][connection] source/common/network/connection_impl.cc:104] [C1] closing data_to_write=0 type=1
[2018-01-19 22:20:45.773][174619][debug][connection] source/common/network/connection_impl.cc:138] [C1] closing socket: 1
[2018-01-19 22:20:45.773][174619][trace][connection] source/common/network/connection_impl.cc:342] [C2] socket event: 1
[2018-01-19 22:20:45.773][174619][trace][connection] source/common/network/connection_impl.cc:382] [C2] read ready
[2018-01-19 22:20:45.773][174619][trace][connection] source/common/network/raw_buffer_socket.cc:23] [C2] read returns: 9
[2018-01-19 22:20:45.773][174619][trace][connection] source/common/network/raw_buffer_socket.cc:23] [C2] read returns: 0
[2018-01-19 22:20:45.773][174619][debug][connection] source/common/network/connection_impl.cc:397] [C2] remote close
[2018-01-19 22:20:45.773][174619][debug][connection] source/common/network/connection_impl.cc:138] [C2] closing socket: 0
[2018-01-19 22:20:45.773][174619][debug][main] source/server/connection_handler_impl.cc:54] [C2] adding to cleanup list
[2018-01-19 22:20:45.773][174619][trace][main] source/common/event/dispatcher_impl.cc:126] item added to deferred deletion list (size=1)
[2018-01-19 22:20:45.773][174619][trace][main] source/common/event/dispatcher_impl.cc:53] clearing deferred deletion list (size=1)
[       OK ] IpVersions/ProxyProtocolTest.Basic/0 (5 ms)
[ RUN      ] IpVersions/ProxyProtocolTest.Basic/1
[2018-01-19 22:20:45.774][174619][debug][connection] source/common/network/connection_impl.cc:499] [C3] connecting to [::1]:53399
[2018-01-19 22:20:45.774][174619][debug][connection] source/common/network/connection_impl.cc:508] [C3] connection in progress
[2018-01-19 22:20:45.774][174619][debug][filter] source/common/filter/listener/proxy_protocol.cc:28] proxy_protocol: New connection accepted
[2018-01-19 22:20:45.774][174619][trace][connection] source/common/network/connection_impl.cc:342] [C3] socket event: 2
[2018-01-19 22:20:45.774][174619][trace][connection] source/common/network/connection_impl.cc:403] [C3] write ready
[2018-01-19 22:20:45.774][174619][debug][connection] source/common/network/connection_impl.cc:413] [C3] connected
[2018-01-19 22:20:45.774][174619][trace][connection] source/common/network/connection_impl.cc:279] [C3] writing 56 bytes
[2018-01-19 22:20:45.774][174619][trace][connection] source/common/network/connection_impl.cc:342] [C3] socket event: 2
[2018-01-19 22:20:45.774][174619][trace][connection] source/common/network/connection_impl.cc:403] [C3] write ready
[2018-01-19 22:20:45.774][174619][trace][connection] source/common/network/raw_buffer_socket.cc:58] [C3] write returns: 56
[2018-01-19 22:20:45.774][174619][debug][connection] source/common/network/connection_impl.cc:104] [C3] closing data_to_write=0 type=1
[2018-01-19 22:20:45.774][174619][debug][connection] source/common/network/connection_impl.cc:138] [C3] closing socket: 1
[2018-01-19 22:20:45.774][174619][debug][filter] source/common/filter/listener/proxy_protocol.cc:152] proxy_protocol: Read 47 bytes (total 47) from socket (found: true)
[2018-01-19 22:20:45.774][174619][debug][main] source/server/connection_handler_impl.cc:193] [C4] new connection
[2018-01-19 22:20:45.774][174619][trace][main] source/common/event/dispatcher_impl.cc:126] item added to deferred deletion list (size=1)
[2018-01-19 22:20:45.774][174619][trace][main] source/common/event/dispatcher_impl.cc:53] clearing deferred deletion list (size=1)
[2018-01-19 22:20:45.774][174619][trace][connection] source/common/network/connection_impl.cc:342] [C4] socket event: 2
[2018-01-19 22:20:45.774][174619][trace][connection] source/common/network/connection_impl.cc:403] [C4] write ready
Terminated: 15
================================================================================
```

lizan commented:
Here is the stacktrace where it hangs, looks like in `disconnect()`.

```
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00007fffd23bdd96 libsystem_kernel.dylib`kevent + 10
    frame #1: 0x00000001088f190d proxy_protocol_test`kq_dispatch + 685
    frame #2: 0x00000001088e79f9 proxy_protocol_test`event_base_loop + 633
    frame #3: 0x0000000107ab3202 proxy_protocol_test`Envoy::Event::DispatcherImpl::run(Envoy::Event::Dispatcher::RunType) + 98
    frame #4: 0x0000000107a2c79c proxy_protocol_test`Envoy::Network::ProxyProtocolTest::disconnect() + 476
    frame #5: 0x0000000107a2cbfd proxy_protocol_test`Envoy::Network::ProxyProtocolTest_BasicV6_Test::TestBody() + 589
    frame #6: 0x0000000108e1d24e proxy_protocol_test`void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 126
    frame #7: 0x0000000108df355b proxy_protocol_test`void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 123
    frame #8: 0x0000000108df3486 proxy_protocol_test`testing::Test::Run() + 198
    frame #9: 0x0000000108df49cd proxy_protocol_test`testing::TestInfo::Run() + 221
    frame #10: 0x0000000108df5f8c proxy_protocol_test`testing::TestCase::Run() + 236
    frame #11: 0x0000000108e0592b proxy_protocol_test`testing::internal::UnitTestImpl::RunAllTests() + 1019
    frame #12: 0x0000000108e1fe9e proxy_protocol_test`bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) + 126
    frame #13: 0x0000000108e052eb proxy_protocol_test`bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) + 123
    frame #14: 0x0000000108e051b8 proxy_protocol_test`testing::UnitTest::Run() + 408
    frame #15: 0x0000000108470941 proxy_protocol_test`RUN_ALL_TESTS() + 17
    frame #16: 0x00000001084707b4 proxy_protocol_test`Envoy::TestRunner::RunTests(int, char**) + 1076
    frame #17: 0x000000010847023f proxy_protocol_test`main + 1983
    frame #18: 0x00007fffd228e235 libdyld.dylib`start + 1
```

jrajahalme commented:
IMO found the problems, will clean up the last few commits after some more testing.

jrajahalme commented:
Proxy protocol tests depended on OS timing w.r.t socket buffering and I/O. Converted tests that did this to run the dispatched until the read filter gets callbacks to make the test runs deterministic. With this I got the proxy protocol test succeed locally on Mac as well as Linux.

jrajahalme commented:
> On Jan 19, 2018, at 21:51, jrajahalme <notifications@github.com> wrote:
> 
> Proxy protocol tests depended on OS timing w.r.t socket buffering and I/O. Converted tests that did this to run the dispatched until the read filter gets callbacks to make the test runs deterministic. With this I got the proxy protocol test succeed locally on Mac as well as Linux.
> 

Should mention that this test likely has been failing or flaky for MAC on the master branch as well for the same reasons.

mattklein123 commented on include/envoy/network/listen_socket.h:
```diff
@@ -34,5 +34,65 @@ class ListenSocket {
+   *        different from the one the socket was initially accepted at. This should only be set
+   *        to 'true' when restoring the original destination address of a connection redirected
+   *        by iptables REDIRECT. The caller is responsible for making sure the new address is
+   *        actually different when passing restored as 'true'.
```
Can you also give an example of when someone would call this function but set `restored` to false? I see multiple examples in this PR but just from reading the interfaces it's not completely clear when one would do that.

jrajahalme replied:
I reverted the meaning of this boolean and renamed it as "hand_off_restored_destinations", should be clearer now, see the new commit.

mattklein123 commented on source/common/filter/listener/proxy_protocol.h:
```diff
@@ -0,0 +1,83 @@
+   */
+  bool readLine(int fd, std::string& s);
+
+  Network::ListenerFilterCallbacks* cb_;
```
initialize to `{}`

jrajahalme replied:
Done.

mattklein123 commented on source/common/filter/listener/proxy_protocol.h:
```diff
@@ -0,0 +1,83 @@
+  size_t buf_off_{};
+
+  // The index in buf_ where the search for '\r\n' should continue from
+  size_t search_index_;
```
not your code but I would move the `{1}` initializer here for consistency.

jrajahalme replied:
OK.

mattklein123 commented on source/common/network/connection_impl.cc:
```diff
@@ -540,9 +478,48 @@ ClientConnectionImpl::ClientConnectionImpl(
+    : ConnectionImpl(dispatcher, std::make_unique<ClientSocketImpl>(remote_address),
+                     std::move(transport_socket), false) {
+  if (source_address != nullptr) {
+    int rc = source_address->bind(fd());
```
nit: const

jrajahalme replied:
OK.

mattklein123 commented on source/common/network/connection_impl.cc:
```diff
@@ -540,9 +478,48 @@ ClientConnectionImpl::ClientConnectionImpl(
+
+void ClientConnectionImpl::connect() {
+  ENVOY_CONN_LOG(debug, "connecting to {}", *this, socket_->remoteAddress()->asString());
+  int rc = socket_->remoteAddress()->connect(fd());
```
nit: const

jrajahalme replied:
Ditto.

mattklein123 commented on source/common/network/listen_socket_impl.h:
```diff
@@ -50,5 +50,55 @@ class UdsListenSocket : public ListenSocketImpl {
+  // Network::ConnectionSocket
+  const Address::InstanceConstSharedPtr& localAddress() const override { return local_address_; }
+  const Address::InstanceConstSharedPtr& remoteAddress() const override { return remote_address_; }
+  void setLocalAddress(const Address::InstanceConstSharedPtr& local_address,
```
can we add some `ASSERT` here that verifies that someone does not violate the invariants that you set out in the function description?

jrajahalme replied:
Good point. Done.

mattklein123 commented on source/server/connection_handler_impl.cc:
```diff
@@ -69,71 +58,138 @@ void ConnectionHandlerImpl::ActiveListener::removeConnection(ActiveConnection& c
+  // Purge sockets that have not progressed to connections. This should only happen when
+  // a listener filter stops iteration and never resumes.
+  while (!sockets_.empty()) {
+    ActiveSocketPtr removed = sockets_.front()->removeFromList(sockets_);
```
Do you have test coverage for this case?

jrajahalme replied:
proxy_protocol_test.cc should exercise this, it has tests for invalid/truncated proxy protocol messages.

mattklein123 commented on source/server/connection_handler_impl.h:
```diff
@@ -130,6 +126,35 @@ class ConnectionHandlerImpl : public Network::ConnectionHandler, NonCopyable {
+
+    ActiveListener& listener_;
+    Network::ConnectionSocketPtr socket_;
+    bool redirected_;
```
const

jrajahalme replied:
OK.

mattklein123 commented on source/server/listener_manager_impl.cc:
```diff
@@ -96,6 +118,30 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, ListenerManag
   // filter chain #1308.
   ASSERT(config.filter_chains().size() >= 1);
 
+  if (config.listener_filters().size()) {
```
nit: `!config.listener_filters().empty()`

jrajahalme replied:
OK.

mattklein123 commented on source/server/listener_manager_impl.cc:
```diff
@@ -96,6 +118,30 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, ListenerManag
+    listener_filter_factories_ =
+        parent_.factory_.createListenerFilterFactoryList(config.listener_filters(), *this);
+  }
+  // Add original dst listener filter if 'use_original_dst' flag is set.
```
Per my comment in the data-plane-api repo, can we deprecate the use_original_dst field in DEPRECATED.md and docs, and push people towards using the filter directly.

jrajahalme replied:
Done.

mattklein123 commented on source/server/listener_manager_impl.cc:
```diff
@@ -96,6 +118,30 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, ListenerManag
+  //                   is implemented, this needs to be run after the filter chain has been
+  //                   selected.
+  if (PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.filter_chains()[0], use_proxy_proto, false)) {
+    auto& factory =
```
I'm wondering if we should even register proxy proto as a factory, given that in the future we aren't going to use it as a listener filter. I don't think it's a big deal either way, but just throwing it out there for thoughts.

jrajahalme replied:
IMO, I would leave the door open to use it as a listener filter as well, in case someone must do filter chain matching after proxy protocol has restored the destination address.

jrajahalme commented:
@mattklein123 Thanks for your thorough review! I'll post commits addressing your comments right after the test run I'm doing locally finishes.

jrajahalme commented:
Will also merge master to resolve the conflict.

mattklein123 commented on source/common/network/listener_impl.h:
```diff
@@ -17,13 +17,14 @@ namespace Network {
 
   Address::InstanceConstSharedPtr local_address_;
   ListenerCallbacks& cb_;
+  bool hand_off_restored_destinations_;
```
const

jrajahalme replied:
I have to learn this someday :-) Btw. this factors in to the discussion about this being plural or not in the other comment..

mattklein123 commented on include/envoy/network/listener.h:
```diff
@@ -43,6 +43,14 @@ class ListenerConfig {
+  /**
+   * @return bool if a connection should be handed off to another Listener after the original
+   *         destination address has been restored. 'true' when 'use_original_dst' flag in listener
+   *         configuration is set, false otherwise. Note that this flag will be removed from the v2
```
can you explicitly use the word "deprecated" in here somewhere so that when we search for deprecated things this comes up?

jrajahalme replied:
Done.

mattklein123 commented on include/envoy/network/listener.h:
```diff
@@ -43,6 +43,14 @@ class ListenerConfig {
+   *         configuration is set, false otherwise. Note that this flag will be removed from the v2
+   *         API.
+   */
+  virtual bool handOffRestoredDestinations() const PURE;
```
Should this be not plural? `handOffRestoredDestination()` ? And same with all variables similarly named?

jrajahalme replied:
A listener accepts many connections, each of which has a destination address, and the listener will either hand all restored destinations off to another listener (if found) or not. To me this calls for a plural, but I'm happy to change it you don't buy this line of reasoning.

mattklein123 replied:
OK makes sense. I guess I would probably name it `handOffRestoredDestinationConnections()`, but I don't feel that strongly about it.

jrajahalme replied:
OK, I'll change this, hold on for a while :-)

jrajahalme commented:
I've been running the starting point of this PR for a couple of months now, but still positively surprised by the code quality & readability increase during the review process! Thanks for keeping a to it for the last two weeks!

mattklein123 commented:
> I've been running the starting point of this PR for a couple of months now, but still positively surprised by the code quality & readability increase during the review process! Thanks for keeping a to it for the last two weeks!

Thank you for being a good sport! :)
